### PR TITLE
Replace synchronized blocks with ReentrantLocks for virtual thread support

### PR DESCRIPTION
**Pull request by omercelikceng**
Replace **synchronized** methods and blocks with **ReentrantLocks** in a few classes in Spring Functions Catalog to improve compatibility with virtual threads. This changes the synchronization mechanism in:

**CassandraConsumerConfiguration.PayloadToMatrixTransformer**
**InMemoryTraceRepository**

The change helps avoid blocking virtual threads when using Spring Functions Catalog in Project Loom environments while maintaining thread safety.

### CODE REVIEW

**Review by artembilan — CHANGES_REQUESTED** [review]
Let’s see if such an action is really necessary is those cases: https://mikemybytes.com/2024/02/28/curiosities-of-java-virtual-threads-pinning-with-synchronized/! 
In other words: if there is really no blocking operation inside that `synchronized`, no reason to change it. Plus it looks like some changes are coming to Java down the road : https://openjdk.org/jeps/491

**Comment by omercelikceng** [comment]
> Let’s see if such an action is really necessary is those cases: https://mikemybytes.com/2024/02/28/curiosities-of-java-virtual-threads-pinning-with-synchronized/! In other words: if there is really no blocking operation inside that `synchronized`, no reason to change it. Plus it looks like some changes are coming to Java down the road : https://openjdk.org/jeps/491

@artembilan Thank you very much for all the valuable information you provided. The links you provided contain really great information.

However, I believe that ReentrantLock offers us more flexibility. Additionally, ReentrantLock provides features like interruptible lock acquisition, timed lock attempts, and finer control with the Condition API, which are not available with synchronized. Of course, the current code doesn't involve very complex operations, but I don't think ReentrantLock would introduce significant overhead either. Therefore, I think we could implement this improvement instead of using synchronized. However, if you disagree, I can close the pull request.

**Comment by artembilan** [comment]
Sure! I’ll give it a second look when back from vacation next week .

File: consumer/spring-cassandra-consumer/src/main/java/org/springframework/cloud/fn/consumer/cassandra/CassandraConsumerConfiguration.java
Comment (artembilan) [inline consumer/spring-cassandra-consumer/src/main/java/org/springframework/cloud/fn/consumer/cassandra/CassandraConsumerConfiguration.java:178]: The `lock()` call is supposed to be outside of the `try` block. See `Lock` JavaDocs.

File: consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java
Comment (artembilan) [inline consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:54]: I don't think we need any locking in setters at all.

File: consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java
Comment (artembilan) [inline consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:78]: Doesn't look like this operation needs any locking at all.

File: consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java
Comment (artembilan) [inline consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:89]: Same here about moving `lock()` call outside of the `try` block.

**Review by artembilan — CHANGES_REQUESTED** [review]
Please, add your name to the `@author` list of the affected classes.

File: consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java
Comment (omercelikceng) [reply consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:54]: Alright, then I’m removing the synchronized keyword as well. Is it correct?

File: consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java
Comment (omercelikceng) [reply consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:78]: Alright, then I’m removing the synchronized keyword as well. Is it correct?

File: consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java
Comment (omercelikceng) [reply consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:54]: @artembilan I think locking is necessary here because the add method on line 86 uses the reverse and capacity values. add method requires locking, as any changes to reverse or capacity while the add method is being called could result in an inconsistent state. Therefore, I believe we should continue using locking.

File: consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java
Comment (omercelikceng) [reply consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:78]: @artembilan I think the same way here as well; we shouldn’t remove the locking from this part either.

File: consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java
Comment (artembilan) [reply consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java]: What you are explaining is correct, but that's not realistic. 
This class is created only once on Application context initialization.
Those setters are called at this point and only once.
Calling them at runtime is really mocking of the lifecycle of the instance of this class.

We should just agree that we are in Spring, so there are some undocumented rules for singleton beans where we just initialize them once. Plus the instance of this class is created by the configuration in this module.
You just need to drop the jar into your classpath - and everything will be auto-configured for you.

Why do we need the extra logic which may not have power in 99% of usage?

File: consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java
Comment (artembilan) [inline consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:54]: You just removed our conversion on the matter, but still left lock in the setter.
🤷

**Review by artembilan — CHANGES_REQUESTED** [review]

File: consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java
Comment (omercelikceng) [reply consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:54]: Once our conversation is finished, I was going to remove them. I'll take care of it right away.

Comment (artembilan) [reply consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:54]: But that conversation has not been finished.
You just didn't wait for my response.
I am the owner of the project, that the final decision is on me.

Please, don't remove any conversation: it will be helpful in the future when some one would try to figure out what was an origin of this or that code.
Like exactly someone would ask why the `synchronized` as removed here and lock not added.

Comment (omercelikceng) [reply consumer/spring-websocket-consumer/src/main/java/org/springframework/cloud/fn/consumer/websocket/trace/InMemoryTraceRepository.java:54]: I just thought I would make another commit after our conversation. Of course, I’ll follow your guidelines. I didn’t mean to do anything wrong, sorry if I caused any confusion.

**Comment by omercelikceng** [comment]
@artembilan I've applied the changes you recommended. And thank you for your guidance.

**Review by artembilan — APPROVED** [review]
All good, but I cannot merge it yet until I cut `5.0.x` branch.
We are planning to release `5.1.0` soon enough.
Thank you!

**Comment by omercelikceng** [comment]
> All good, but I cannot merge it yet until I cut `5.0.x` branch. We are planning to release `5.1.0` soon enough. Thank you!

Thank you very much for everything. Best regards.

**Comment by artembilan** [comment]
@omercelikceng ,
I've changed my mind.
Since it looks like we are going to have `5.0.2` soon enough, this is a good candidate to be included there having its only internal nature.
Thank you and looking forward for more!

**Pull request merged**
